File: 8/800s/826/MostProfitAssigningWork.java
/*
https://leetcode.com/problems/most-profit-assigning-work/description/?envType=daily-question&envId=2024-06-18
*/

class CustomComparator implements Comparator<Pair>{

    @Override
    public int compare(Pair A, Pair B){
        return Integer.compare(A.getFirst(), B.getFirst());
    }
}

class Pair{

    public int first;

    public int second;

    Pair(int first, int second){
        this.first=first;
        this.second=second;
    }

    public int getFirst(){
        return this.first;
    }

    public int getSecond(){
        return this.second;
    }
}

class Solution {
    public int maxProfitAssignment(int[] difficulty, int[] profit, int[] worker) {
        ArrayList<Pair>difficultyProfitCombination=new ArrayList<>();
        ArrayList<Pair>profitDifficultyCombination=new ArrayList<>();
        int difficultyProfitLength=difficulty.length, resultantMaxProfit=0;
        for(int i=0; i<difficultyProfitLength; ++i){
            difficultyProfitCombination.add(new Pair(difficulty[i], profit[i]));
            profitDifficultyCombination.add(new Pair(profit[i], difficulty[i]));
        }
        Collections.sort(difficultyProfitCombination, new CustomComparator());
        Collections.sort(profitDifficultyCombination, new CustomComparator());
        int workerLength=worker.length;
        for(int i=0; i<workerLength; ++i){
            int currentWorkerLimit=worker[i], traversalIndexInFirstList=0;
            boolean isAnyDifficultyProfitCombinationValid=false;
            while(traversalIndexInFirstList<=difficultyProfitLength-1){
                if(difficultyProfitCombination.get(traversalIndexInFirstList).getFirst()<=currentWorkerLimit){
                    if(!isAnyDifficultyProfitCombinationValid){
                        isAnyDifficultyProfitCombinationValid=true;
                    }
                    traversalIndexInFirstList++;
                } else break;
            }
            int currentCollectedProfit=0;
            if(isAnyDifficultyProfitCombinationValid){
                if(traversalIndexInFirstList>=1) traversalIndexInFirstList--;
                currentCollectedProfit=difficultyProfitCombination.get(traversalIndexInFirstList).getSecond();
            }
            int traversalIndexInSecondList=0;
            while(traversalIndexInSecondList<=difficultyProfitLength-1){
                if(profitDifficultyCombination.get(traversalIndexInSecondList).getSecond()<=currentWorkerLimit){
                    currentCollectedProfit=Math.max(currentCollectedProfit, profitDifficultyCombination.get(traversalIndexInSecondList).getFirst());
                }
                traversalIndexInSecondList++;
            }
            resultantMaxProfit+=currentCollectedProfit;
        }
        return resultantMaxProfit;
    }
}

File: 8/800s/826/MostProfitAssigningWork.cpp
/*
https://leetcode.com/problems/most-profit-assigning-work/description/?envType=daily-question&envId=2024-06-18
*/

class Solution {
public:
    int maxProfitAssignment(vector<int>& difficulty, vector<int>& profit, vector<int>& worker) {
        vector<pair<int, int>>comb1, comb2;
        int sz=difficulty.size(), res=0;
        for(int i=0; i<sz; ++i){
            comb1.push_back({difficulty[i], profit[i]});
            comb2.push_back({profit[i], difficulty[i]});
        }
        sort(comb1.begin(), comb1.end());
        sort(comb2.begin(), comb2.end());
        for(auto &vals: worker){
            int idx=0;
            bool flag=false;
            while(idx<=sz-1){
                if(comb1[idx].first<=vals){
                    if(!flag) flag=true;
                    idx++;
                } else break;
            }
            int prttochk=0;
            if(flag){
                if(idx>=1) idx--;
                prttochk=comb1[idx].second;
            }
            int idx2=0;
            while(idx2<=sz-1){
                if(comb2[idx2].second<=vals){
                    prttochk=max(prttochk, comb2[idx2].first);
                }
                idx2++;
            }
            res+=prttochk;
        }
        return res;
    }
};